### ISSUE TITLE
Better JS debugging experience

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#9035]
ATM, debugging JS code is very annoying because of the amount of different variables that are generated by the compiler. I often have like tens of _gXX or other making it difficult to look at "real" local variables values.

Also, "real" locals are often renamed, making it difficult to track down which is the actual runtime name of the locale.

I think we could improve things significantly here by:
- avoiding renaming locals as much as possible - we currently do some kind of "global scope rename" for JS, but we could be smarter than that
- optimize reusage of generated variable names, so we don't have many of them in the scope
- eventually use ES6 "let" ? I'm not sure if this allows to have properly scoped variable in debugger, having them leave the variables list as soon as we escape the current block

**Comment by RealyUniqueName** [comment]
Sourcemaps spec has var name mapping covered. We just need to support it.

**Comment by nadako** [comment]
Yeah we gotta figure out the `names` field. However there are some specific things that can be fixed regarding naming. E.g.
```haxe
		function f(a, b) {
			var c = 3;
		}

		function f2(a, b) {
			var c = 3;
		}

		function f3(a, b) {
			var c = 3;
		}
```
will generate:
```js
	var f = function(a,b) {
		var c = 3;
	};
	var f2 = function(a1,b1) {
		var c1 = 3;
	};
	var f3 = function(a2,b2) {
		var c2 = 3;
	};
};
```
which makes no sense and hurts debugging.

**Comment by Simn** [comment]
I guess this requires "flushing" the renaming operation at the end of each nested function.

**Comment by ncannasse** [comment]
Any update on this ? Is #9045 still the way to go ?

**Comment by ncannasse** [comment]
@RealyUniqueName I think we need first to only do the renaming that are strictly necessary, so one looking at JS output can "read it" the way it was written - sometimes you have to debug JS source code without source maps. And then once this is done, it's still possible to add source maps for the other renames (and hopefully it allows filtering of temp vars, which is another big issue).

**Comment by RealyUniqueName** [comment]
Resolved with #9304 and #9280
